tool/tsh/tsh.go

```diff
@@ -2249,20 +2249,99 @@ func showDatabases(w io.Writer, clusterFlag string, databases []types.Database,
 	return nil
 }
 
-func serializeDatabases(databases []types.Database, format string) (string, error) {
+func serializeDatabases(databases []types.Database, format string, roleSet services.RoleSet) (string, error) {
 	if databases == nil {
 		databases = []types.Database{}
 	}
+
+	if roleSet != nil {
+		dbsWithUsers, err := getDatabasesWithUsers(databases, roleSet)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		return serializeDatabasesWithUsers(dbsWithUsers, format)
+	}
+
 	var out []byte
 	var err error
-	if format == teleport.JSON {
+	switch {
+	case format == teleport.JSON:
 		out, err = utils.FastMarshalIndent(databases, "", "  ")
-	} else {
+	default:
 		out, err = yaml.Marshal(databases)
 	}
 	return string(out), trace.Wrap(err)
 }
 
+type dbUsers struct {
+	Allowed []string `json:"allowed,omitempty"`
+	Denied  []string `json:"denied,omitempty"`
+}
+
+type databaseWithUsers struct {
+	// *DatabaseV3 is used instead of types.Database because we want the db fields marshaled to JSON inline.
+	// An embedded interface (like types.Database) does not inline when marshaled to JSON.
+	*types.DatabaseV3
+	DBUsers *dbUsers `json:"db_users"`
+}
+
+func getDBUsers(db types.Database, roles services.RoleSet) *dbUsers {
+	users := roles.EnumerateDatabaseUsers(db)
+	var denied []string
+	allowed := users.Allowed()
+	if users.WildcardAllowed() {
+		// start the list with *.
+		allowed = append([]string{types.Wildcard}, allowed...)
+		// only include denied users if the wildcard is allowed.
+		denied = append(denied, users.Denied()...)
+	}
+	return &dbUsers{
+		Allowed: allowed,
+		Denied:  denied,
+	}
+}
+
+func newDatabaseWithUsers(db types.Database, roles services.RoleSet) (*databaseWithUsers, error) {
+	dbWithUsers := &databaseWithUsers{
+		DBUsers: getDBUsers(db, roles),
+	}
+	switch db := db.(type) {
+	case *types.DatabaseV3:
+		dbWithUsers.DatabaseV3 = db
+	default:
+		return nil, trace.BadParameter("unrecognized database type %T", db)
+	}
+	return dbWithUsers, nil
+}
+
+func getDatabasesWithUsers(databases types.Databases, roles services.RoleSet) ([]*databaseWithUsers, error) {
+	var dbsWithUsers []*databaseWithUsers
+	for _, db := range databases {
+		dbWithUsers, err := newDatabaseWithUsers(db, roles)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		dbsWithUsers = append(dbsWithUsers, dbWithUsers)
+	}
+	return dbsWithUsers, nil
+}
+
+func serializeDatabasesWithUsers(dbsWithUsers []*databaseWithUsers, format string) (string, error) {
+	// initialize to an empty slice so we don't print `null` in JSON/YAML if the cluster has no databases.
+	if dbsWithUsers == nil {
+		dbsWithUsers = []*databaseWithUsers{}
+	}
```

Comment:
This function only differes from `serializeDatabases` end flow by this call. 
Instead of returning raw string we can make a helper function something like like 
```
func getDatabasePrintObject(databases, ...) any, error {
       if roleSet != nil {
		dbsWithUsers, err := getDatabasesWithUsers(databases, roleSet)
		if err != nil {
			return "", trace.Wrap(err)
		}
		return dbsWithUsers
	}
       return databases
}
```
and call `getDatabasePrintObject` function without adding duplication